Restore phandles from binary representations

ukleinek opened the pull request:
device trees generated with options `-@` and `-L` contain information about the label names used originally and which values are phandle values. This information can be reused when compiling back to dts format to make the result much more human readable.

Also dtdiff is adapted to make use of this to generate smaller diffs without hunks e.g. for a different phandle allocation.

This superseeds pull request https://github.com/dgibson/dtc/pull/93.

ukleinek committed:
Restore labels from __symbols__ node

If the input has a __symbols__ node, restore the named labels for the
respective nodes.

Signed-off-by: Uwe Kleine-König <u.kleine-koenig@baylibre.com>

ukleinek committed:
Restore phandle references from __local_fixups__ node

The __local_fixups__ node contains information about phandles. Parse it
to improve the result when decompiling a device tree blob.

Note that this is essential to do before the __local_fixups__ node is
removed in generate_local_fixups_tree() because otherwise the
information contained in that node is lost.

Fixes: 915daadbb62d ("Start with empty __local_fixups__ and __fixups__ nodes")
Signed-off-by: Uwe Kleine-König <u.kleine-koenig@baylibre.com>

ukleinek commented:
A possible followup improvement would be to restore phandles from `__fixup__` .

ukleinek commented:
Just a quick headsup: I found a dtbo file where my code crashes with `-@ -L`. So don't merge yet please.

dgibson commented on dtc.c:
```diff
@@ -336,8 +336,10 @@ int main(int argc, char *argv[])
-	if (generate_symbols)
+	if (generate_symbols) {
 		generate_label_tree(dti, "__symbols__", true);
+		generate_labels_from_tree(dti, "__symbols__");
```
I'd not immediately sure if it make sense to only do this when `generate_symbols` is set.

ukleinek replied:
Maybe do it unconditionally and remove the `__symbols__` node if `generate_symbols` is set?

dgibson replied:
Yes, I think that makes sense.

dgibson commented on livetree.c:
```diff
@@ -1048,6 +1048,28 @@ static void generate_local_fixups_tree_internal(struct dt_info *dti,
+		if (labeled_node)
+			add_label(&labeled_node->labels, p->name);
+		else if (quiet < 1)
+			fprintf(stderr, "Warning: Path %s referenced in %s missing",
```
I'd suggest suggest an explicit "referenced in symbol %s" to make it clearer what's going on without context.

ukleinek replied:
I guess you expect that `p->name` is passed then for your `%s`? I think mentioning `name` is still a good idea, so I suggest to make this
```
fprintf(stderr, "Warning: Path %s referenced in property %s/%s missing",
        p->val.val, name, p->name);
```

dgibson commented on dtc.c:
```diff
@@ -344,6 +344,7 @@ int main(int argc, char *argv[])
 	if (generate_fixups) {
 		generate_fixups_tree(dti, "__fixups__");
 		generate_local_fixups_tree(dti, "__local_fixups__");
+		local_fixup_phandles(dti, "__local_fixups__");
```
Again, not sure if it only makes sense to do this with `-L`.

dgibson replied:
As with `__symbols__` I think it probably makes sense to do this unconditionally, then strip it out of `-L` is not given.

dgibson commented on livetree.c:
```diff
@@ -3,6 +3,8 @@
  */
 
+#include <libfdt/libfdt.h>
+
```
Oof.  We don't currently use libfdt within dtc, which is sort-of deliberate (they provide a cross-check to each other).  There's a case to be made to drop that policy and instead use libfdt for much of the flat tree logic within dtc.  However, I'd really prefer to do that as an overall change, using it for as much as we can, rather than adding small pieces of libfdt usage ad-hoc.

..also, I can't actually see what you're using this for.

ukleinek replied:
Maybe this is only a development relict, will try without.

ukleinek replied:
FTR: Just dropping the #include doesn't work. Have to take a look at a more appropriate time.

dgibson commented on livetree.c:
```diff
@@ -1093,3 +1095,59 @@ void generate_local_fixups_tree(struct dt_info *dti, const char *name)
+		}
+
+		/*
+		 * Each property in the __local_fixup__ tree is a concatination
```
Nits: 'concatenation' and `__local_fixups__` .

dgibson commented on livetree.c:
```diff
@@ -1093,3 +1095,59 @@ void generate_local_fixups_tree(struct dt_info *dti, const char *name)
+		}
+
+		for (i = 0; i < lfp->val.len; i += sizeof(fdt32_t))
+			add_phandle_marker(dti, p, fdt32_ld((fdt32_t *)(lfp->val.val + i)));
```
Not a big deal, but I'd prefer to use `i += 1` with the loop limit decreased correspondingly.  Then you can cast `lfp->val.val` to `fdt32_t *` once and just index into it with `i`.

dgibson commented on livetree.c:
```diff
@@ -1093,3 +1095,59 @@ void generate_local_fixups_tree(struct dt_info *dti, const char *name)
+
+		if (!subnode) {
+			if (quiet < 1)
+				fprintf(stderr, "Warning: subnode %s in %s referenced in __local_fixups__ missing\n",
```
For properties you do need to give the property name and path separately.  However for subnodes you might as well just stick them together "node %s/%s referenced in __local_fixups__ ...".

dgibson commented on treesource.c:
```diff
@@ -161,6 +163,36 @@ static void add_string_markers(struct property *prop)
+			m->ref = refn->fullpath;
+	} else if (quiet < 1) {
+		fprintf(stderr, "Warning: node referenced by phandle 0x%x in property %s not found\n",
+			phandle, prop->name);
```
I think you do need to consider `__fixups__` as well.  Without that, this error will be triggered if you're decompiling a dtbo with any external phandle references.

dgibson commented on dtdiff:
```diff
@@ -17,7 +17,7 @@ source_and_sort () {
 		IFORMAT=dts
 		;;
-	    *.dtb)
+	    *.dtb|*.dtbo)
```
Even something this simple needs some sort of commit message: why is this change desirable?  (*I* can see why, but someone reading the commit log in the future might not).

dgibson reviewed (changes requested):
I think this is a great idea.  The implementation looks mostly sound but there are a few minor issues that need addressing.

dgibson commented:
> A possible followup improvement would be to restore phandles from `__fixup__` .

I think you do need to implement this.  For one thing processing but `__local_fixups__` but not `__fixups__` would just be surprising behaviour to a user, I think.  Secondly, it could cause some bogus errors in some cases (see detailed comments).

ukleinek commented:
Do you have an idea for the unaligned phandle problem?

dgibson commented:
> Do you have an idea for the unaligned phandle problem?

Sorry, I'm not sure exactly what problem you mean.

dgibson commented:
> > > Do you have an idea for the unaligned phandle problem?
> > 
> > 
> > Sorry, I'm not sure exactly what problem you mean.
> 
> I mean the crash reported in [#151 (comment)](https://github.com/dgibson/dtc/pull/151#issuecomment-2434383526) with the follow up in the next comment.

Ah, right, sorry.

I think the basic problem is that previously the only case where we'd have phandle markers during decompile is with `-I dts -O dts`, in which case we'd also have type markers helping us figure out how to print each part of each property.  The new code is putting in phandle markers, but without type markers.  We should handle that, of course, but looks like there are some edge cases.

More specifically, it looks like `write_propval()` calls `guess_value_type()` to figure out a missing type only once for the whole property.  Having a phandle marker does tell us that those specific 4 bytes should be in integer `< ... >` context.  But we'll need to separately guess the type for each chunk between phandles.  I suspect that might be irritatingly fiddly, but I'm afraid it's going to be a prerequisite for implementing this.

ukleinek commented:
This addresses the crash mentioned in https://github.com/dgibson/dtc/pull/151#issuecomment-2434383526 and a few suggestions by @dgibson (which I marked as resolved). Still missing is restoring of phandles from `__fixups__`.

ukleinek committed:
Restore phandle references from __fixups__ node

The __fixups__ node contains information about labels. Parse its
properties to create phandle markers which improve the resulting dts
when decompiling a device tree blob.

Note that this is essential to do before the __fixups__ node is
removed in generate_fixups_tree() because otherwise the information
contained in that node is lost.

Fixes: 915daadbb62d ("Start with empty __local_fixups__ and __fixups__ nodes")
Signed-off-by: Uwe Kleine-König <u.kleine-koenig@baylibre.com>

ukleinek committed:
tests: Check for duplication in __local_fixups__ and __fixups__

Typically the info contained in __local_fixups__ and __fixups__ is
autogenerated from phandles used in the tree. However before commit
915daadbb62d ("Start with empty __local_fixups__ and __fixups__ nodes")
compiling a dts that has both the __local_fixups__ and/or __fixups__ and
phandles that result in entries in these nodes, the fixups were
duplicated:

	$ cat test.dts
	/dts-v1/;
	/plugin/;

	/ {
		important: node {
			property = <&somenode>, "string";
			phandle = <0x01>;
			self = <&important>;
		};

		__fixups__ {
			somenode = "/node:property:0";
		};

		__local_fixups__ {
			node {
				self = <0x00>;
			};
		};

		__symbols__ {
			important = "/node";
		};
	};

	$ dtc -v
	Version: DTC 1.7.2

	$ dtc -O dts test2.dts
	/dts-v1/;

	/ {

		important: node {
			property = <&somenode>, "string";
			phandle = <0x01>;
			self = <&important>;
		};

		__fixups__ {
			somenode = "/node:property:0", "/node:property:0";
		};

		__local_fixups__ {

			node {
				self = <0x00>, <0x00>;
			};
		};

		__symbols__ {
			important = "/node";
		};
	};

Add a test that ensures this issue isn't reintroduced later.

Signed-off-by: Uwe Kleine-König <u.kleine-koenig@baylibre.com>

ukleinek committed:
New helper to add markers

The add_marker() function is used to create a new marker and add it at
the right spot to the relevant marker list. Use it in the
add_string_markers() helper (which gets slightly quicker by it).

Signed-off-by: Uwe Kleine-König <u.kleine-koenig@baylibre.com>

ukleinek committed:
Improve type guessing when compiling to dts format

In the presence of (non-type) markers guess the type of each chunk
between markers individually instead of only once for the whole
property.

Note that this only gets relevant with the next few commits that restore
labels and phandles. Note further that this rework is necessary with
these further changes, because phandle markers are currently not
considered for type guessing and so a phandle at an offset that isn't a
multiple of 4 triggers an assertion if the property was guessed to have
type TYPE_UINT32.

Now that guess_value_type() is only called for data chunks without
markers, the function can be simplified a bit.

Signed-off-by: Uwe Kleine-König <u.kleine-koenig@baylibre.com>

ukleinek commented:
OK, this got bigger now than I anticipated and also contains a few fixes, but I'm happy that I did it now and I can get it out of my head and happily use it.

Quick demo:
```
$ cat test.dts
/dts-v1/;
/plugin/;

/ {
	tralal = <&somelabel>, "text making the",
		<&somelabel>, "some more text";

	tralala = <&somelabel>, "text making the next label unaligned",
		<&somelabel>, "some more text";

	somelabel: somenode {
		property = <&nonexisting>;
	};
};
$ dtc -@ test.dts > test.dtb
```
Decompiling this dtb with dtc 1.7.2 gives you:
```
$ /usr/bin/dtc test.dtb
/dts-v1/;

/ {
	tralal = [00 00 00 01 74 65 78 74 20 6d 61 6b 69 6e 67 20 74 68 65 00 00 00 00 01 73 6f 6d 65 20 6d 6f 72 65 20 74 65 78 74 00];
	tralala = <0x01 0x74657874 0x206d616b 0x696e6720 0x74686520 0x6e657874 0x206c6162 0x656c2075 0x6e616c69 0x676e6564 0x00 0x1736f6d 0x65206d6f 0x72652074 0x65787400>;

	somenode {
		property = <0xffffffff>;
		phandle = <0x01>;
	};

	__symbols__ {
		somelabel = "/somenode";
	};

	__fixups__ {
		nonexisting = "/somenode:property:0";
	};

	__local_fixups__ {
		tralal = <0x00 0x14>;
		tralala = <0x00 0x29>;
	};
};
```
With the changes from this PR applied it gets:
```
$ dtc test.dtb
/dts-v1/;
/plugin/;

/ {
	tralal = <&somelabel>, "text making the", <&somelabel>, "some more text";
	tralala = <&somelabel>, "text making the next label unaligned", <&somelabel>, "some more text";

	somelabel: somenode {
		property = <&nonexisting>;
		phandle = <0x01>;
	};

	__symbols__ {
		somelabel = "/somenode";
	};

	__fixups__ {
		nonexisting = "/somenode:property:0";
	};

	__local_fixups__ {
		tralal = <0x00 0x14>;
		tralala = <0x00 0x29>;
	};
};
```
and if `test.dtb` was compiled without `-@` it uses `&{/somenode}` instead of `&somelabel`.

ukleinek commented on treesource.c:
```diff
@@ -178,7 +177,9 @@ static enum markertype guess_value_type(struct property *prop)
 	for_each_marker_of_type(m, LABEL) {
-		if ((m->offset > 0) && (prop->val.val[m->offset - 1] != '\0'))
+		if (m->offset < offset || m->offset >= offset + len)
+			continue;
```
Either I'm not following you, or you're wrong. If there is a label in the middle of a data chunk, I consider that a boundary and guess the types of the two chunks separately. IMHO that makes sense because a label (or a phandle) is very likely a boundary for a type change.

ukleinek commented on treesource.c:
```diff
@@ -139,16 +139,16 @@ static const char *delim_end[] = {
 
 		m = xmalloc(sizeof(*m));
-		m->offset = l;
+		m->offset = offset + l;
```
This is right in the new code. I think it was already correct before, but let's not waste time on that. Please convince you that the current version is fine.

ukleinek commented:
BTW, I didn't implement stripping out the `__fixups__`, `__local_fixups__` and `__symbols__` nodes. Mostly because it doesn't seem to be right for `-I dts -O dts` compilations. Instead the two fixup nodes are removed before they are generated to make sure that the information isn't duplicated.

dgibson commented on treesource.c:
```diff
@@ -347,7 +347,10 @@ void dt_to_source(FILE *f, struct dt_info *dti)
+	fprintf(f, "/dts-v1/;\n");
+	if (any_fixup_tree(dti, dti->dt))
+		fprintf(f, "/plugin/;\n");
+	fprintf(f, "\n");
```
I think this is correct, but I seems like a very indirect way of diong it.  We're taking the `__fixups__` node, distributing across the tree as a bunch of fixup markers, then scanning the entire tree for fixup markers to mark it as `/plugin/`.

I think a clear way to do this would be to set the `DTSF_PLUGIN` flag on dtb input if there is a (non empty) `__fixups__` or `__local_fixups__` node, then on dts output output the `/plugin/` tag based on that flag.

ukleinek replied:
I went with testing just for the existence of `__fixups__` and `__local_fixups__`. I think the effort to check for non-emptiness (which has a different semantic for the two subtrees) is overengineering that isn't worth being invested.

dgibson commented on tests/over-determined.dts:
```diff
@@ -0,0 +1,23 @@
+/dts-v1/;
```
This commit definitely needs a proper commit message.  What's the rationale for this test?

ukleinek replied:
I don't understand this concern. IMHO the commit message is fine. What are you missing?

dgibson commented on treesource.c:
```diff
@@ -139,26 +139,49 @@ static const char *delim_end[] = {
 
+/*
+ * The invariants in the marker list are:
+ *  - offsets are monotonically rising
```
Maybe "non-strictly monotonically increasing", since adjacent offsets can be equal.

dgibson commented on treesource.c:
```diff
@@ -139,26 +139,49 @@ static const char *delim_end[] = {
+ * The invariants in the marker list are:
+ *  - offsets are monotonically rising
+ *  - for a single offset there is at most one type marker
+ *  - for a single offset there is at most one non-type marker
```
This one isn't true.
```
prop = &somenode, <&othernode>;
```
Will have the `REF_PATH` and `REF_PHANDLE` markers at the same offset, until the `REF_PATH` marker is resolved (before that it's essentially marking a region of zero length).

dgibson reviewed (commented):
I've applied the first three patches, because they're sensible fixes independent of the rest of the series.  I've made some comments on the next few, but I haven't completed a full review yet.

ukleinek commented:
> I've applied the first three patches, because they're sensible fixes independent of the rest of the series. I've made some comments on the next few, but I haven't completed a full review yet.

The topmost commit is also simple and orthogonal to the rest of the series. That one might be eligible for fast tracking, too.
I'll look into your other feedback later.

ukleinek committed:
Emit /plugin/ when compiling to .dts with DTSF_PLUGIN set

This fixes `dtc -I dts -O dts` to make the file a plugin if the source
file is one.

Signed-off-by: Uwe Kleine-König <u.kleine-koenig@baylibre.com>

ukleinek committed:
Set DTSF_PLUGIN if needed when compiling from dtb

The need for the plugin flag is determined by the existence of __fixups__
or __local_fixups__.

This is a bit simplifying because if __fixups__ or __local_fixups__
exist but don't have properties, the plugin flag isn't needed. But in
practise the test should be good enough such that this corner case
doesn't matter.

Signed-off-by: Uwe Kleine-König <u.kleine-koenig@baylibre.com>

ukleinek commented:
I think I addressed all feedback. The only open discussion is if `__fixups__`, `__local_fixups__` and `__symbols__` should be dropped and under which conditions.

dgibson commented:
The first two patches looked fine and I was going to apply.. but the second one breaks a test case:
```
dtc -I dtb -O dts -o overlay_overlay_decompile.test.dts overlay_overlay.test.dtb:	PASS
dtc -I dts -O dtb -o overlay_overlay_decompile.test.dtb overlay_overlay_decompile.test.dts:PASS
dtbs_equal_ordered overlay_overlay.test.dtb overlay_overlay_decompile.test.dtb:	Starting testcase "./dtbs_equal_ordered", pid 98779
FAIL	Tag mismatch (1 != 2) at (672, 672)
```

It looks like the decompiled version is entirely missing the `__fixups__` and `__local_fixups__` which should be there.  I'm guessing this eventually gets fixed by the later patches which will reconstruct the necessary source level references to remake the fixups.  However, we don't like to break bisection like that, so that patch might have to be moved later to where it won't break things.

dgibson commented on treesource.c:
```diff
@@ -139,26 +139,48 @@ static const char *delim_end[] = {
+ *  - for a single offset that has both a type marker and non-type markers, the
+ *    type marker appears before the others.
+ */
+static struct marker **add_marker(struct marker **mi,
```
Is there a reason to take a general `struct marker **` here instead of a `struct property *`?

dgibson commented on treesource.c:
```diff
@@ -199,30 +196,49 @@ static enum markertype guess_value_type(struct property *prop)
+		if ((*m)->type == REF_PHANDLE) {
+			m = add_marker(m, TYPE_UINT32, offset, NULL);
+			offset += 4;
+		}
```
Although... that will require revising your invariant of at most one type marker per offset.

dgibson commented on livetree.c:
```diff
@@ -1090,6 +1090,107 @@ void generate_fixups_tree(struct dt_info *dti, const char *name)
+
+			if (!p && quiet < 1)
+				fprintf(stderr, "Warning: Label %s references non-existing property %s in node %s",
+					fp->name, n->fullpath, propname);
```
Another `continue;`?

dgibson commented on livetree.c:
```diff
@@ -1090,6 +1090,107 @@ void generate_fixups_tree(struct dt_info *dti, const char *name)
+					fprintf(stderr,
+						"Warning: Label %s contains invalid offset for property %s in node %s",
+						fp->name, p->name, n->fullpath);
+				break;
```
I think `continue;` would make more sense - if one fixup is bogus, you still parse the others.  The fixups are badly constructed either way, so I guess `break;` is ok, but as a general rule we try to keep going when we can.

Of couse using `continue;` here, as with the other places I suggested does mean changing from a `while` to a `for` and moving the update of `fl` and `fv` into the `for` clause.

dgibson commented on treesource.c:
```diff
@@ -173,6 +173,12 @@ static struct marker **add_marker(struct marker **mi,
 	return &nm->next;
 }
 
+void property_add_marker(struct property *prop,
```
This belongs in `livetree.c`

ukleinek replied:
Is this true with `add_marker` in `treesource.c`?

dgibson commented on tests/run_tests.sh:
```diff
@@ -570,6 +570,10 @@ dtc_tests () {
 
+    run_dtc_test -I dts -O dts -o over-determined-once.dts "$SRCDIR/over-determined.dts"
+    run_dtc_test -I dts -O dts -o over-determined-twice.dts "over-determined-once.dts"
+    run_wrap_test cmp over-determined-once.dts over-determined-twice.dts
```
I generally mistrust directly comparing dts files, since it can pick up meaningless format differences.  But I guess both files you're comparing here are generated by the same dtc version, so it's probably ok.

ukleinek commented:
> The first two patches looked fine and I was going to apply.. but the second one breaks a test case:
> 
> ```
> dtc -I dtb -O dts -o overlay_overlay_decompile.test.dts overlay_overlay.test.dtb:	PASS
> dtc -I dts -O dtb -o overlay_overlay_decompile.test.dtb overlay_overlay_decompile.test.dts:PASS
> dtbs_equal_ordered overlay_overlay.test.dtb overlay_overlay_decompile.test.dtb:	Starting testcase "./dtbs_equal_ordered", pid 98779
> FAIL	Tag mismatch (1 != 2) at (672, 672)
> ```
> 
> It looks like the decompiled version is entirely missing the `__fixups__` and `__local_fixups__` which should be there. I'm guessing this eventually gets fixed by the later patches which will reconstruct the necessary source level references to remake the fixups. However, we don't like to break bisection like that, so that patch might have to be moved later to where it won't break things.

This commit uncovers a problem that we already have since commit 915daadbb62d ("Start with empty __local_fixups__ and __fixups__ nodes"):

    $ cat test.dts 
    /dts-v1/;
    /plugin/;
    
    / {
    	node {
    		property = <0xffffffff>, "string";
    	};
    	
    	__fixups__ {
    		somenode = "/node:property:0";
    	};
    };

    $ dtc -I dts -O dts test.dts 
    /dts-v1/;
    
    / {
    
    	node {
    		property = <0xffffffff>, "string";
    	};
    };

I'll think about a fix.

ukleinek commented on tests/run_tests.sh:
```diff
@@ -570,6 +570,10 @@ dtc_tests () {
     run_dtc_test -I dts -O dtb -o dtc_sized_cells.test.dtb "$SRCDIR/sized_cells.dts"
     run_test sized_cells dtc_sized_cells.test.dtb
 
+    run_dtc_test -I dts -O dts -o over-determined-once.dts "$SRCDIR/over-determined.dts"
```
I fixed that locally here, will be better after the next force-push.

ukleinek commented on treesource.c:
```diff
@@ -183,6 +183,31 @@ static void add_string_markers(struct property *prop, unsigned int offset, int l
+		else
+			ref = refn->fullpath;
+	} else {
+		if (quiet < 1)
```
Fixed locally here, will collect a few more changes before pushing again.

ukleinek commented on livetree.c:
```diff
@@ -1090,6 +1090,107 @@ void generate_fixups_tree(struct dt_info *dti, const char *name)
+			n = get_node_by_path(dti->dt, fv);
+			if (!n && quiet < 1)
+				fprintf(stderr, "Warning: Label %s references non-existing node %s",
+					fp->name, fv);
```
There is a `break;` a few lines further down. `*(propname - 1) = ':'` cannot be done before the error message is emitted.

dgibson replied:
Oh, sorry, I missed that.  Ugh, this is why I kind of hate temporarily mangling the data in place, not that I can really see a better way in this instance.

ukleinek commented on livetree.c:
```diff
@@ -1090,6 +1090,107 @@ void generate_fixups_tree(struct dt_info *dti, const char *name)
+		unsigned int fl = fp->val.len;
+
+		while (fl) {
+			char *fend = memchr(fv, 0, fl);
```
I don't understand that comment.

dgibson replied:
Sorry, I started writing something in the wrong place, and for some reason github was being weird and didn't let me delete it properly.

dgibson commented on treesource.c:
```diff
@@ -173,23 +173,20 @@ static struct marker **add_marker(struct marker **mi,
 
 	for (l = strlen(p) + 1; l < len; l += strlen(p + l) + 1)
-		mi = add_marker(mi, TYPE_STRING, l, NULL);
+		mi = add_marker(mi, TYPE_STRING, offset + l, NULL);
```
Ok, fair.

dgibson commented:
> This commit uncovers a problem that we already have since commit [915daad](https://github.com/dgibson/dtc/commit/915daadbb62d68bc49c5a14a360d1c39e0131f97) ("Start with empty **local_fixups** and **fixups** nodes"):
> 
> ```
> $ cat test.dts 
> /dts-v1/;
> /plugin/;
> 
> / {
> 	node {
> 		property = <0xffffffff>, "string";
> 	};
> 	
> 	__fixups__ {
> 		somenode = "/node:property:0";
> 	};
> };
> 
> $ dtc -I dts -O dts test.dts 
> /dts-v1/;
> 
> / {
> 
> 	node {
> 		property = <0xffffffff>, "string";
> 	};
> };
> ```
> 
> I'll think about a fix.

Ouch, ok.  I missed that nasty side effect of [915daad](https://github.com/dgibson/dtc/commit/915daadbb62d68bc49c5a14a360d1c39e0131f97).  Unfortunately I don't think the fix proposed in the latest version is quite sufficient.

It's an important principle in dtc's design that it should generally be safe and lossless to convert dtb -> dts -> dtb.  Likewise it should be "lossless" to convert dts -> dts, in the sense that going directly dts -> dtb and going dts -> dts -> dtb should result in the same output.  This is exactly what those test cases are designed for.

This is supposed to be true even if the input tree contains garbage.  It needs the basic structure to be parseable, of course, but if the properties contain things that are semantically malformed, they should be passed through unchanged.  915daad broke that.

Your latest series partially fixes this, but not completely: it will still simply delete entries it doesn't understand in the input tree.  It might also re-order entries that it does understand.  I can see two ways to fully fix this:

1) Change fixup generation, so that instead of unconditionally adding the new fixup, it first checks to see if a matching fixup is already in `__fixups__` or `__local_fixups__`.  With that in place, 915daad can be reverted.  That fix could be applied immediately, then your code to understand the existing fixups could be done on top of that.  Alas, that is kind of messy and a fair bit of work.

2) This approach might be a bit easier... but I'm not sure it gets every case correct.  Instead of wiping `__fixups__` clean as a single step, you remove entries one at a time as you parse them into fixups in the live tree (so you now know they'll be re-generated).  This could still result in re-orderings which isn't great, but maybe acceptable.  I'm also not sure if it handles all breakages from 915daad that don't involve a dtb -> dts stage.

ukleinek commented:
FTR: I took this patch set to the mailing list as I prefer this workflow after this PR got unwieldy.
v1: https://lore.kernel.org/devicetree-compiler/cover.1755692822.git.u.kleine-koenig@baylibre.com
v2: https://lore.kernel.org/devicetree-compiler/20250919092912.663304-8-u.kleine-koenig@baylibre.com

dgibson commented:
Equivalent changes merged via mail.  Closing.
